Number parser verifies if its already parsed

rhiroshi opened the pull request:
Hello.
This PR will fix a problem where the number parser was clearing the true separator when trying to delocalize a number that didnt need to delocalize.  
My problem was that in i would delocalize a number like '123.456,78' to '123456.78'. Then ransack's sort_link would create a url with this number already delocalized (eg: params[:search][:number]=123456.78) then, if i clicked that sort, it would be delocalized again to '12345678'. 

Thanks

rhiroshi committed:
Number parser verifies if its already parsed

daniel-rikowski commented:
I'm not sure it is possible to reliably detect if a string was already delocalized.

If I read your code correctly, a German localized integer like `12.345` (delimiter == `.`, separator == `,`) would normally be delocalized into `12345`, but with your change  it wouldn't be touched at all. Subsequentially it would be parsed incorrectly as `12.345` (written as a Ruby float) because the delimiter is interpreted as separator.

It looks like your specific problem arises from writing back the delocalized string to `params`. If you can avoid mutating `params`, e.g. by chaining the delocalize call with the require/permit call and only use its return value, you don't need to detect if a string already is delocalized. 

Something like this (not tested):

```ruby
def search_params
  return {} if params[:search].blank?
  params.require(:search).permit(:number).delocalize(number: :number)
end

def index
  MyModel.all.ransack(search_params)
end
```

Your actual searching code (controller) would use `search_params` with the delocalized value and the call to `sort_link` (view) would use the original values of `params`. At least in principle that's what I do in my applications.

rhiroshi commented:
I see. That wouldn't work then. I was thinking only about decimals not integers. Will close it. 
The problem is that when i pass Model.ransack(params[:search]), Ransack will copy the params and pass to sort_link. I have no access to what ransack writes on sort_link so ransack will write a delocalized params.

daniel-rikowski commented:
It seems I underestimated the problem.

That ransack calll uses params to both perform the query - which needs delocalized values - and to provide the context to sort_link - which needs the original values.

I'm not sure how to tackle this...

clemens commented:
From how I understand the problem, a potential approach would be to check for the Rails `utf8` parameter that the form sends (and the link shouldn't) and depending on this either call or don't call `delocalize`.

Pick up on Daniel's code (and also untested):

``` ruby
def search_params
  return {} if params[:search].blank?

  search_params = params.require(:search).permit(:number)
  search_params = search_params.delocalize(number: :number) if params[:utf8].present?

  search_params
end
```

Caveat: It seems that Rails 6 doesn't inject this parameter by default anymore (see https://github.com/rails/rails/pull/32125). So you'd either have to explicitly enable it or use your own hidden field (e.g. `hidden_field_tag :delocalize, 1` => `if params[:delocalize] == '1'` in the controller).

daniel-rikowski commented:
Another way would be to use a custom Ransack predicate:

```ruby
Ransack.configure do |config|
  config.add_predicate 'deloc_eq', # or any other name
                       arel_predicate: 'eq',
                       formatter: proc {|v| Delocalize::Parsers::Number.new.parse(v) }
end
```

Then you can use `number_deloc_eq: '12.345,67'` and it should work with user input directly, without any additional delocalize call. 

In case you want to skip delocalization - for example when you programmatically build a Ransack query - you can fallback to the `eq` predicate, e.g. `number_eq: '12345.67'`


See https://github.com/activerecord-hackery/ransack/issues/34#issuecomment-2504138 for a similar discussion on date searches.

rhiroshi commented:
Thanks both of you for your help and time. 

Daniel's last comment did solve my problem. 

As it was a custom ransack, i added `Delocalize::Parsers::Number.new.parse(v)` to the ransacker and removed the delocalization from params.delocalize. 

sort_link is not building links with the delocalized params anymore!

Thanks again.
